packages/cardpay-sdk/sdk/scheduled-payment-module.ts

```diff
@@ -653,48 +671,194 @@ export default class ScheduledPaymentModule {
       )
     )[0];
 
-    return signature;
+    return { nonce, estimate, payload, signature };
   }
 
-  private async generateSchedulePaymentTxParams(
+  private async schedulePaymentOnChainAndUpdateCrank(
+    hubRootUrl: string,
+    authToken: string,
+    scheduledPaymentId: string,
     safeAddress: string,
     moduleAddress: string,
     gasTokenAddress: string,
     spHash: string
   ) {
-    let contract = new this.web3.eth.Contract(ScheduledPaymentABI as AbiItem[], moduleAddress);
-    let payload = await contract.methods.schedulePayment(spHash).encodeABI();
+    let txnParams = await this.generateSchedulePaymentTxParams(safeAddress, moduleAddress, gasTokenAddress, spHash);
```

Comment:
I think we can put this part on the `scheduledPaymentOnChain` function as well.

Reply:
You mean like this? (it's already there): https://github.com/cardstack/cardstack/pull/3257/files#diff-d85a97b0da51ca637fc08ca1b5f782d0d6bdc949ccf414465ab0236d607db756R891

Reply:
Yes, so I think this line can be removed?

Reply:
I want to get the signature and the rest of the params before submitting the data to the hub. So in case anything fails with `generateSchedulePaymentTxParams`, we won't be saving anything to the hub. That's why I'm generating this data earlier and passing it in.

Reply:
But I don't see you could achieve that by putting `generateSchedulePaymentTxParams` here. If you want to make sure the `generateSchedulePaymentTxParams` work before submitting to the hub, you should put it in the `schedulePayment` before post payment data to the hub. And I think it's unnecessary because you already had the error catch to delete the data to the hub if the `schedulePaymentOnChain` failed.

Reply:
Oh yeah I see now, I should have placed `generateSchedulePaymentTxParams` call higher in the call chain. For defensive purposes I still think it's more robust to attempt that call before the POST rather than rely on the error handler cleanup